hypershift-operator/metrics.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/prometheus/client_golang/prometheus"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+type hypershiftMetrics struct {
+	// clusterCreationTime is the time it takes between cluster creation until the first
+	// version got successfully rolled out. Technically this is a const, but using a gauge
+	// means we do not have to track what we already reported and can just call Set
+	// repeatedly with the same value.
+	clusterCreationTime *prometheus.GaugeVec
+
+	client crclient.Client
+
+	log logr.Logger
+}
+
+func (m *hypershiftMetrics) Start(ctx context.Context) error {
+	ticker := time.NewTicker(30 * time.Second)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			if err := m.collect(ctx); err != nil {
+				m.log.Error(err, "failed to collect metrics")
+			}
+		}
+	}
+}
+
+func (m *hypershiftMetrics) collect(ctx context.Context) error {
+	var clusters hyperv1.HostedClusterList
+	if err := m.client.List(ctx, &clusters); err != nil {
+		return fmt.Errorf("failed to list hostedclusters: %w", err)
+	}
+
+	for _, cluster := range clusters.Items {
+		if cluster.Status.Version == nil || len(cluster.Status.Version.History) == 0 {
+			continue
+		}
+		completionTime := cluster.Status.Version.History[len(cluster.Status.Version.History)-1].CompletionTime
+		if completionTime == nil {
+			continue
+		}
+		m.clusterCreationTime.WithLabelValues(cluster.Namespace + "/" + cluster.Name).Set(completionTime.Sub(cluster.CreationTimestamp.Time).Seconds())
```

Comment:
Over time with recycled namespaces/names would this mess up the series by overlapping distinct (in time) cluster metrics? Using UUID would be one way to do it if we care about the identity of the metric...

Reply:
Speaking of identity, I wonder what it means to us here. Do we care about individual clusters in the series, e.g. for granular debugging, or the higher level trends over time compared to SLOs? For example we could consider a histogram instead if we can define some target SLOs in terms of buckets. Any thoughts along those lines?

Reply:
Thinking some more, histogram may be an oversimplification because not all clusters are the same re: an SLO, depending on all sorts of factors (HA mode, node pools, platform, infra config, etc. etc.).

Reply:
And re: UUID, if we want the identities, I'm not sure if that will be enough to correlate any of our existing metrics which I think only capture the namespace and name (in discrete labels)... if we want the namespace and name, should they be separate labels?